images/prow-tests/Dockerfile

```diff
@@ -41,32 +41,47 @@ RUN apt-get install -y uuid-runtime  # for uuidgen
 RUN apt-get install -y rubygems      # for mdl
 RUN apt-get install -y shellcheck    # for presubmit
 
-# Install go at 1.13 same as how it's installed in kubekins-e2e image, reference code here:
-# https://github.com/kubernetes/test-infra/blob/1e9b5dc3de4b268aab57c9c48120aa3dcf096bc6/images/kubekins-e2e/Dockerfile#L64
-ENV GO_TARBALL "go1.13.linux-amd64.tar.gz"
-RUN rm -rf /usr/local/go && \
-    wget -q "https://storage.googleapis.com/golang/${GO_TARBALL}" && \
-    tar xzf "${GO_TARBALL}" -C /usr/local && \
-    rm "${GO_TARBALL}"
+# Extra tools through gem
+RUN gem install mixlib-config -v 2.2.4  # required because ruby is 2.1
+RUN gem install mdl -v 0.5  # required because later version of mdl requires ruby >= 2.4
+
+# Install go 1.12, 1.13, and 1.14 using https://github.com/moovweb/gvm
+# To use this, in Prow jobs set env var GO_VERSION="go1.14" (or go1.13, go1.12, etc)
+# (Obviously the version of Go must be installed on the image to be available -- we don't want to be downloading and installing it over and over again)
+# Missing prerequisite:
+RUN apt-get install -y bison
+# Install the tool:
+RUN curl -s -S -L https://raw.githubusercontent.com/moovweb/gvm/master/binscripts/gvm-installer | bash
+# gvm requires one to "source /root/.gvm/scripts/gvm" after installing
+#  but in Dockerfile, each RUN is its own shell, so source'd in-RUN env vars are not propagated
+# So have created "my-gvm.sh" to source the above then run gvm
+COPY images/prow-tests/my-gvm.sh /usr/local/bin
+# Install our versions of Go
+RUN my-gvm.sh install go1.12 --prefer-binary
+RUN my-gvm.sh install go1.13 --prefer-binary
+RUN my-gvm.sh install go1.14 --prefer-binary
```

Comment:
I cannot remember clearly, but does `gvm` support installing minor go versions? Is there a way to update to a patch release like `go1.14.1`?

Reply:
yes:
```
root@fe6a5d65896e:/workspace# gvm install go1.13.1 --prefer-binary
Installing go1.13.1 from binary source
```

Reply:
Then let's install the minor version `go1.14.1` so the environment is more deterministic and more clear when we really need an update?

Reply:
It does... but we've never needed it before and then you also have to update all the Prow jobs to use the new GO_VERSION... but I'm fine with that if you are

Reply:
There's also a Go 1.14.2 so we should probably jump straight to it

Reply:
Hmm I tried 1.14.2 on Mac but it told me the binary was not found. But if works in the container I agree we can directly jump to it.

Reply:
And it seems we don't need to specify the minor version when using it, `gvm use go1.14` will just work.

Reply:
Interactively I installed go1.11.2:
`gvm install go1.11.2 --prefer-binary`
then
`gvm use go1.11` and it said:
`Now using version go1.11.2`

so you are right!

Installing go1.14.2 worked:
```
root@435c7b9d5d89:/workspace# gvm install go1.14.2 --prefer-binary
Installing go1.14.2 from binary source
```
So I'll install 1.14.2 in the image and check the use command works, then upload.

Reply:
```
root@f9b34d9b819c:/workspace# gvm use go1.14
Now using version go1.14.2
root@f9b34d9b819c:/workspace# go version
go version go1.14.2 linux/amd64
root@f9b34d9b819c:/workspace# gvm use go1.13
Now using version go1.13.10
root@f9b34d9b819c:/workspace# go version
go version go1.13.10 linux/amd64
root@f9b34d9b819c:/workspace# 
```